ruby/ql/lib/codeql/ruby/security/UnsafeShellCommandConstructionQuery.qll

```diff
@@ -0,0 +1,35 @@
+/**
+ * Provides a taint tracking configuration for reasoning about shell command
+ * constructed from library input vulnerabilities
+ *
+ * Note, for performance reasons: only import this file if `Configuration` is needed,
+ * otherwise `UnsafeShellCommandConstructionCustomizations` should be imported instead.
+ */
+
+import codeql.ruby.DataFlow
+import UnsafeShellCommandConstructionCustomizations::UnsafeShellCommandConstruction
+private import codeql.ruby.TaintTracking
+private import CommandInjectionCustomizations::CommandInjection as CommandInjection
+private import codeql.ruby.dataflow.BarrierGuards
+
+/**
+ * A taint-tracking configuration for detecting shell command constructed from library input vulnerabilities.
+ */
+class Configuration extends TaintTracking::Configuration {
+  Configuration() { this = "UnsafeShellCommandConstruction" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    node instanceof CommandInjection::Sanitizer or // using all sanitizers from `rb/command-injection`
+    node instanceof StringConstCompareBarrier or
+    node instanceof StringConstArrayInclusionCallBarrier
+  }
+
+  // override to require the path doesn't have unmatched return steps
```

Comment:
Consider this test case: 

```Ruby
class Foobar2
  def id(x)
    return x
  end 

  def thisIsSafe()
    IO.popen("echo #{id('foo')}", "w") # <- We risk an FP here
  end    
end
```

When unmatched return steps are included we can have a path that starts with the `x` in `id`, and returns out of `id` into the `popen` call in `thisIsSafe`, which causes an obvious FP to be reported.  
The return-step from `id` to `thisIsSafe` doesn't have a matching call, so that's why I can block it.